src/api/spec/controllers/webui/projects/public_key_controller_spec.rb

```diff
@@ -27,7 +27,8 @@
     context 'with a project that has no public key' do
       let(:keyinfo_response) { '<keyinfo />' }
 
-      it { expect(response.status).to eq(404) }
+      it { is_expected.to redirect_to(project_show_path(project)) }
+      it { expect(flash[:error]).not_to be_nil }
```

Comment:
I would prefer that one: `it { is_expected.to set_flash[:error].to include('not', 'public key') }`

But still I could change the error message to something completely wrong which contains `not` and `public key` and your test would pass. :sob: 

> I don't want to end up with a code base where the tests are so specific that they don't allow refactoring and basically become a carbon copy of the code

You can refactor everything, but the error message should be the same, but this makes sense because the message is part of the functionality and shouldn't be changed when refactoring.